### PR TITLE
Async watching and uploading including tarring experiment

### PR DESCRIPTION
**Pull request by AydinHassan**
😱 

TODO:
- [x] Tests for `Delete`
- [x] Tests for `FileExistsInContainer`
- [x] Tests for `WorkflowTest`

### CODE REVIEW

File: src/Files.php [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            return sprintf('/var/www/%s', $source);
+        });
+
+        $makeDirectoriesCommand = sprintf(
+            'docker exec %s mkdir -p %s',
+            $container,
+            $files->map(toMap('dirname'))->map(toMap('escapeshellarg'))->implode(' ')
+        );
+
+        $this->runCommand($makeDirectoriesCommand, function () use ($currentDir, $container, $sources, $destinations) {
```
Comment (AydinHassan) [inline src/Files.php]: Callback hell 😱

File: test/fixtures/test-env/.docker/local.env [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Copy to local.env and store on local machine
```
Comment (mikeymike) [inline test/fixtures/test-env/.docker/local.env]: from what I can see we don't need this right?

File: test/fixtures/test-env/app.php.dockerfile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+FROM php:7.0-fpm
+MAINTAINER Michael Woodward <michael@wearejh.com>
+
+ARG BUILD_ENV=dev
```
Comment (mikeymike) [inline test/fixtures/test-env/app.php.dockerfile]: sack these two off 👍

File: test/fixtures/test-env/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+version: '2'
+
+volumes:
```
Comment (mikeymike) [inline test/fixtures/test-env/docker-compose.yml]: do we need the volumes ?

File: test/fixtures/test-env/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    app-env:
+
+services:
+    php:
+        container_name: m2-php
+        image: workflow-test
+        build:
+            context: ./
+            dockerfile: app.php.dockerfile
+        volumes:
```
Comment (mikeymike) [inline test/fixtures/test-env/docker-compose.yml]: can probbaly remove these too

**Review by mikeymike — APPROVED** [review]
Looks awesome dude, few comments around the docker stuff but the rest I can't fault

**Pull request merged**